File: Solutions/Problem4.ts
var shortestPalindrome = function(s:string):string {
    let pattern:string = s.split('').reverse().join('') //The string but in reverse
    let patternIndex:number = 0 //An index so we can 
    let string:string = s+"*"+pattern;//This is the string that will help us to find the lps
    let lps:number[] = [] //Longest Propper Prefix, used in the KPM algorithm. 
    let aux:number;//Just an auxiliar for the index
    lps[0] = 0
    for(let i = 1; i < string.length; i++)//This loop fill the lps array
    {
        while(string[i] != string[patternIndex] && patternIndex > 0)
        {
            aux = patternIndex
            patternIndex = lps[aux-1]
        }
        if(string[i] == string[patternIndex])
        {
            patternIndex += 1
            
        }
        lps[i] = patternIndex
    }
    let palindrome = pattern.slice(0, pattern.length-patternIndex) + s;//Took the prefix and put it in front of the original string.
    return palindrome
};

File: Solutions/Problem4.java
class Solution {
    public String shortestPalindrome(String s) {
        String pattern = "";
        for(int i = 0; i < s.length(); i++)
        {
            pattern = s.charAt(i) + pattern;
        }
        int patternIndex = 0;
        String string = s + "*" + pattern;
        int[] lps = new int[string.length()];
        lps[0] = 0;
        for(int i = 1; i < string.length(); i++)
        {
            while((string.charAt(i) != string.charAt(patternIndex)) && (patternIndex > 0))
            {
                patternIndex = lps[patternIndex-1];
            }
            if(string.charAt(i) == string.charAt(patternIndex))
            {
                patternIndex ++;
            }
            lps[i] = patternIndex;
        }
        String front = "";
        for(int i = 0; i < (pattern.length()-patternIndex); i++)
        {
            front = front + pattern.charAt(i);
        }
        String palindrome = front + s;
        System.out.println(pattern);
        return pattern;
    }
}